File: downloadFiles.py
import aiohttp
import asyncio
import aiofiles
import os
import time

file_path = 'urls.txt'
download_directory = './downloaded_sds_files'
os.makedirs(download_directory, exist_ok=True)

# optimize connection pool settings
MAX_CONNECTIONS = 100

async def download_pdf(session, url, download_directory):
    try:
        async with session.get(url) as response:
            response.raise_for_status()
            
            filename = url.split('/')[-1]
            file_path = os.path.join(download_directory, filename)
            
            # read and write the content asynchronously
            content = await response.read()
            async with aiofiles.open(file_path, 'wb') as file:
                await file.write(content)
            
            print(f"Downloaded: {filename}")
    except Exception as e:
        print(f"Failed to download {url}. Error: {e}")

async def download_pdfs_from_file(file_path, download_directory):
    try:
        with open(file_path, 'r') as file:
            urls = [url.strip() for url in file if url.strip()]

        # start benchmarking
        start_time = time.time()
        print("Download started...")

        # configure the connector to allow more connections and tune timeouts
        connector = aiohttp.TCPConnector(limit=MAX_CONNECTIONS)
        timeout = aiohttp.ClientTimeout(total=None, connect=60, sock_read=60)

        # use a single session for all requests to improve performance
        async with aiohttp.ClientSession(connector=connector, timeout=timeout) as session:
            tasks = [download_pdf(session, url, download_directory) for url in urls]
            await asyncio.gather(*tasks)

        # end benchmarking
        end_time = time.time()
        duration = end_time - start_time
        print(f"Download completed in {duration:.2f} seconds.")

    except FileNotFoundError:
        print(f"File not found: {file_path}")

# running the main async function
if __name__ == '__main__':
    asyncio.run(download_pdfs_from_file(file_path, download_directory))

File: downloadFiles.js
import fs from "fs-extra";
import axios from "axios";
import path from "path";
import readline from "readline";
import pLimit from "p-limit";

const downloadDir = "./downloads";

// create download directory if it does not exist
fs.ensureDirSync(downloadDir);

// read URLs from urls.txt
const urls = fs.readFileSync("urls.txt", "utf-8").split("\n").filter(Boolean);
const totalFiles = urls.length;

// time benchmarking
const startTime = Date.now();

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const updateProgress = (completed, total) => {
  const percent = ((completed / total) * 100).toFixed(2);
  readline.cursorTo(process.stdout, 0);
  process.stdout.write(`Progress: ${completed}/${total} (${percent}%)`);
};

// the main function
const downloadFile = async (url) => {
  try {
    const fileName = path.basename(url);
    const filePath = path.join(downloadDir, fileName);

    const writer = fs.createWriteStream(filePath);
    const response = await axios({
      url,
      method: "GET",
      responseType: "stream",
    });

    response.data.pipe(writer);

    return new Promise((resolve, reject) => {
      writer.on("finish", resolve);
      writer.on("error", reject);
    });
  } catch (error) {
    console.error(`Error downloading ${url}: ${error.message}`);
  }
};

// concurrency limit
const limit = pLimit(100); // adjust this number to find a sweet spot

// function to handle multiple concurrent downloads
const downloadFiles = async (urls) => {
  let completedDownloads = 0;

  const tasks = urls.map((url) =>
    limit(async () => {
      await downloadFile(url);
      completedDownloads++;
      updateProgress(completedDownloads, totalFiles);
    })
  );

  await Promise.all(tasks);
};

downloadFiles(urls)
  .then(() => {
    const endTime = Date.now();
    const duration = ((endTime - startTime) / 1000).toFixed(2);
    console.log("\nAll files have been downloaded.");
    console.log(`Time taken: ${duration} seconds`);
    rl.close();
  })
  .catch((error) => {
    console.error(`Error during download: ${error.message}`);
    rl.close();
  });
